### PR TITLE
Stats: move countries to redux.

### PR DESCRIPTION
**Pull request by timmyc**
This branch is still in-progress and will be blocked by probably a few more forthcoming branches.

TODO:
- [x] Fix stats normalizer tests
- [x] Add `UpgradeNudge` logic back
- [x] Create selector for state-tree based CSV data - #10509
- [x] Update `<DownloadCSV/>` to use state data - #10510

@youknowriad if you have time and want to check out this branch so far, that would be great.

__To Test__
- Open a site stats day page for a site with visits data
- Note the countries module loads as expected and displays the same data as production
- Try to make sure the loading indicators also show as expected, it helps to have the API sandboxed which slows things down
- Click on the 'View All' link or the title bar to visit the Countries Summary page.

### CODE REVIEW

**Comment by jblz** [comment]
<3

**Comment by timmyc** [comment]
Okay all of the requisite blocking PR's have been merged in, and this has been rebased.  Ready for 👀

File: client/my-sites/stats/stats-download-csv/index.jsx [hunk 1]
```diff
@@ -87,7 +86,7 @@ const connectComponent = connect( ( state, ownProps ) => {
 		isLoading = dataList.isLoading();
 	} else {
 		data = getSiteStatsCSVData( state, siteId, statType, query );
-		isLoading = isRequestingSiteStatsForQuery( state, siteId, statType, query );
+		isLoading = isRequestingSiteStatsForQuery( state, siteId, statType, query ) && ! data;
```
Comment (jblz) [inline client/my-sites/stats/stats-download-csv/index.jsx]: I'm not sure the `isLoading` is working as expected to disable the button.

On a large site, I was able to visit the "Year" summary view & click the "Download Data as CSV" button before there was usable data in place (i.e. immediately & before the chart and viz had activated).

I was served an empty CSV file.

Well, mostly empty. The file contained a single BOM / Byte order mark / U+FEFF character. That same character is present in a data-populated CSV, so probably not related to this change.

**Comment by timmyc** [comment]
Thanks @jblz 

> try and reproduce being able to pull an "empty" CSV & prevent that. I'm never really seeing that CSV download button as disabled at all

Yes the disabled feature was added in #10510 - let me see what is happening there.

**Comment by timmyc** [comment]
@jblz fixed up those items in eb430d5 - I just keep `isLoading = true` of the `data` is empty. This results in the button being disabled if there is no data to download, which seems like the right thing to do.

**Comment by timmyc** [comment]
@youknowriad if you have a chance to review this that would be great too 🙌

File: client/my-sites/stats/stats-download-csv/index.jsx [hunk 2]
```diff
@@ -87,7 +86,7 @@ const connectComponent = connect( ( state, ownProps ) => {
 		isLoading = dataList.isLoading();
 	} else {
 		data = getSiteStatsCSVData( state, siteId, statType, query );
-		isLoading = isRequestingSiteStatsForQuery( state, siteId, statType, query );
+		isLoading = isRequestingSiteStatsForQuery( state, siteId, statType, query ) && ! data.length;
```
Comment (jblz) [inline client/my-sites/stats/stats-download-csv/index.jsx]: I'd personally move the `data.length` check for this to the render function since `data` is already in the props & `isLoading` isn't semantically correct if the result is just empty.

Something like [this](https://gist.github.com/anonymous/049d7a0f73b44c5e416d7fb22d0177ac/178fb0438dccba1871cee7bb8d37d1a86434cb77)

File: client/my-sites/stats/geochart/index.jsx [hunk 3]
```diff
@@ -1,26 +1,40 @@
...
+		data: PropTypes.array,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.visualizationsLoaded = false;
+		this.visualization = null;
+		this.tick = () => {
+			this.timer = setTimeout( this.loadVisualizations, 1000 );
+		};
```
Comment (youknowriad) [inline client/my-sites/stats/geochart/index.jsx]: What if you use class properties to avoid creating a constructor?

File: client/my-sites/stats/geochart/index.jsx [hunk 4]
```diff
@@ -50,75 +64,99 @@ export default React.createClass( {
 			this.resize.cancel();
 		}
 		window.removeEventListener( 'resize', this.resize );
-	},
+	}
 
-	recordEvent() {
+	recordEvent = () => {
 		analytics.ga.recordEvent( 'Stats', 'Clicked Country on Map' );
```
Comment (youknowriad) [inline client/my-sites/stats/geochart/index.jsx]: Since we're refactoring this component, could we update this to use `recordGoogleEvent` from `state/analytics/actions`

File: client/my-sites/stats/geochart/index.jsx [hunk 5]
```diff
@@ -50,75 +64,99 @@ export default React.createClass( {
...
+
+		if ( ! data || ! data.length ) {
+			return;
+		}
+
+		const mapData = map( data, ( country ) => {
+			return [ country.label, country.value ];
+		} );
+		mapData.unshift( [
+			this.props.translate( 'Country' ).toString(),
```
Comment (youknowriad) [inline client/my-sites/stats/geochart/index.jsx]: Could we add `translate` to the props destructuring ?

File: client/my-sites/stats/geochart/index.jsx [hunk 6]
```diff
@@ -50,75 +64,99 @@ export default React.createClass( {
...
-		return <div ref="chart" className="stats-geochart" />;
+		const { siteId, statType, query, data } = this.props;
+		const isLoading = ! data;
+		const classes = classNames( 'stats-geochart', {
+			'is-loading': isLoading
+		} );
+		return (
+			<div>
+				<div ref="chart" className={ classes } />
+				{ siteId && statType && <QuerySiteStats statType={ statType } siteId={ siteId } query={ query } /> }
```
Comment (youknowriad) [inline client/my-sites/stats/geochart/index.jsx]: Curious to know why did you add these checks `siteId && statType &&`?

**Comment by timmyc** [comment]
> Not easy to reproduce, but sometimes I see a double loading like this:

The double loader is expected, the top is the placeholder for the visualization, the lower is for list.  I'm going to rebase this and merge after a few more tweaks.  I'll grab the other updates you suggested @youknowriad in the follow-up PR that adds in "Last 7/30/Quarter/All Time"

File: client/my-sites/stats/geochart/index.jsx (hunk 6 above)
Comment (timmyc) [inline client/my-sites/stats/geochart/index.jsx]: Both are required, but `statType` will never be `null`.  Habits.  I'll remove the `statType` check.

**Comment by timmyc** [comment]
@youknowriad I went ahead and did _all the things_ in dcb584e.  Would you mind one last 👀 

Worth noting, just this morning I started to get warnings in my console around no map key from Google.  Fun.

**Comment by jblz** [comment]
> Worth noting, just this morning I started to get warnings in my console around no map key from Google. Fun.

@timmyc I noticed the same & made sure it was happening on master too (it is).

I took a gander at the [docs](https://developers.google.com/chart/interactive/docs/gallery/geochart#region-geocharts) & looks like the demonstrated method of initializing the API has diverged from [our method](https://github.com/Automattic/wp-calypso/blob/c3688a4d31ac63620519761a997fb35cd111eab9/client/my-sites/stats/geochart/index.jsx#L27). 

There's a good chance that `https://www.gstatic.com/charts/loader.js` supports different arguments or has different defaults than `https://www.google.com/jsapi`.

**Comment by jblz** [comment]
Interesting. From https://developers.google.com/chart/interactive/docs/basic_load_libs#limitations:

> For Geochart and Map Chart, you must load both the old library loader and the new library loader.

```
<script type="text/javascript" src="https://www.gstatic.com/charts/loader.js"></script>
<script type="text/javascript" src="https://www.google.com/jsapi"></script>
```
